### PR TITLE
fix grad thrashing of shape analysis

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#40939 fix grad thrashing of shape analysis**
* #40938 shape analysis fix for default dtype'

Previously, when we would do shape analysis by running the op with representative inputs, we would always set the grad property to false. This led to a wrong static analysis when we would create differentiable subgraphs, and propagate shapes without also propagating requires_grad, and then uninline them.

Differential Revision: [D22394676](https://our.internmc.facebook.com/intern/diff/D22394676)

### CODE REVIEW

File: test/test_jit.py [hunk 1]
```diff
@@ -6802,6 +6802,16 @@ def foo(s: float):
...
 
+    def test_shape_analysis_grad_property(self):
+        @torch.jit.script
+        def foo(x):
+            return torch.sub(x, torch.tanh(x))
+
+        torch._C._jit_pass_complete_shape_analysis(foo.graph, (torch.tensor([0.39]),), False)
+
+        # requires_grad property shouldn't be accidentally set by shape analysis
+        self.assertTrue(foo.graph.findNode("aten::sub").output().requiresGrad() is None)
```
Comment (Krovatkin) [inline test/test_jit.py:6813]: couldn't we actually infer in this particular case that `aten::sub` doesn't need `requires_gard`?

**Review by Krovatkin — CHANGES_REQUESTED** [review]

File: test/test_jit.py (hunk 1 above)
Comment (eellison) [reply test/test_jit.py:6813]: No, we haven’t done grad analysis, we don’t know if it requires grad or not

File: torch/csrc/jit/passes/shape_analysis.cpp [hunk 2]
```diff
@@ -421,8 +421,14 @@ class ShapePropagator {
       // some ops may have mixed tensor/primitive outputs
       // for primitives, we don't need to change the type because it is already
       // its most constrained form.
-      if (stack[i].isTensor())
-        node->outputs()[i]->inferTypeFrom(stack[i].toTensor());
+      auto tensor_type = node->outputs()[i]->type()->cast<TensorType>();
```
[the remark this answers is not in the document]
Comment (eellison) [reply torch/csrc/jit/passes/shape_analysis.cpp:424]: Yes, that’s what we’re doing with tensortype::create(stack[i].toTensor())

[the remark this answers is not in the document]
Comment (eellison) [reply torch/csrc/jit/passes/shape_analysis.cpp:424]: Infer type from calls tensortype::create

File: test/test_jit.py (hunk 1 above)
Comment (Krovatkin) [reply test/test_jit.py:6813]: oh got it! thanks!

**Review by Krovatkin — APPROVED** [review]
:shipit:

File: torch/csrc/jit/passes/shape_analysis.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (eellison) [reply torch/csrc/jit/passes/shape_analysis.cpp:424]: Don’t think we can call withRequiresGrad bc that would get rid of grad analysis that has set it to false. I think we just want to preserve whatever our previous grad inference was

**Pull request closed without merging**